Python/Product/Analysis/LanguageServer/ServerBase.cs

```diff
@@ -0,0 +1,232 @@
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<Reference[]> GotoDefinition(TextDocumentPositionParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<Reference[]> FindReferences(ReferencesParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<DocumentHighlight[]> DocumentHighlight(TextDocumentPositionParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<SymbolInformation[]> DocumentSymbol(DocumentSymbolParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<Command[]> CodeAction(CodeActionParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<CodeLens[]> CodeLens(TextDocumentPositionParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<CodeLens> CodeLensResolve(CodeLens item) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<DocumentLink[]> DocumentLink(DocumentLinkParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<DocumentLink> DocumentLinkResolve(DocumentLink item) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<TextEdit[]> DocumentFormatting(DocumentFormattingParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<TextEdit[]> DocumentRangeFormatting(DocumentRangeFormattingParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<TextEdit[]> DocumentOnTypeFormatting(DocumentOnTypeFormattingParams @params) {
+            throw new NotImplementedException();
+        }
+
+        public virtual async Task<WorkspaceEdit> Rename(RenameParams @params) {
+            throw new NotImplementedException();
+        }
+
+        #endregion
+
+        #region Server Requests
+
+        public event EventHandler<ShowMessageEventArgs> OnShowMessage;
+
+        protected void ShowMessage(MessageType type, string message) => OnShowMessage?.Invoke(this, new ShowMessageEventArgs { type = type, message = message });
+
+        public event EventHandler<LogMessageEventArgs> OnLogMessage;
+
+        protected void LogMessage(MessageType type, string message) => OnLogMessage?.Invoke(this, new LogMessageEventArgs { type = type, message = message });
+
+        public event EventHandler<TelemetryEventArgs> OnTelemetry;
+
+        protected void Telemetry(TelemetryEventArgs e) => OnTelemetry?.Invoke(this, e);
+
+        public event EventHandler<RegisterCapabilityEventArgs> OnRegisterCapability;
+        protected Task RegisterCapability(RegistrationParams @params) {
```

Comment:
`RegisterCapability`, `UnregisterCapability` and related events aren't used. Is it for the future functionality?

Reply:
Part of the LSP spec